Vision Transparency

Jolijn2020 opened the pull request:
I added the boolean variable "visible", "transparent" and "solid" to Boundary.
Using these, it is possible to only create rays for certain kinds of intersections.

The vision now can contain multiple rays with the same angle, in the case of transparent boundary intersections before a non-transparent boundary intersection.

The usage of vision has been adjusted to reflect this too (both in the agents and the tests).

The same principle can be used for using visual markers, set visible and transparent to true, and solid to false.
We can now also easily make glass and portals visible in the same way. I haven't done it yet though, to not mess with the current agents.

Jolijn2020 committed:
added visible- and solid- variables to BoundaryImp and made a VisibleTransparentBoundary class

Jolijn2020 committed:
made GraphicsEngine return multiple rays if there are visible but transparent boundaries

Jolijn2020 committed:
adjusted use of rays for AcoAgent, DirectionFollowAgent and the GameEngine (calculating legal moves)

now it first checks if the type it sees is solid or not

Jolijn2020 committed:
adjusted use of rays for WallFollowAgent

now it first checks if the type it sees is solid or not

Jolijn2020 committed:
small bug fix and moved print-statements into if(DEBUG)-condition

Jolijn2020 committed:
fixed new usage of vision in AcoAgent

Jolijn2020 committed:
deleted commented out code

Jolijn2020 committed:
fix in WF-agent

Jolijn2020 committed:
made WF-tests independent of eachother (I think)

S010MON commented on src/main/java/app/model/agents/AgentImp.java:
```diff
@@ -33,6 +33,9 @@ public class AgentImp implements Agent
     protected final boolean DRAW_HEARD = false;
+    @Getter protected boolean solid;
+    @Getter protected boolean visible;
+    @Getter protected boolean transparent;
```
Rather than initialising these three variables three times over in three different constructors, why not just initialise them at  creation time?
```java
 @Getter protected boolean solid = true;
 ```

Jolijn2020 replied:
I'll set some default values yeah, since mose are solid, visible and not transparent.

Jolijn2020 committed:
Merge remote-tracking branch 'origin/master' into make_target_visible

# Conflicts:
#	src/main/java/app/model/agents/WallFollow/WallFollowAgent.java
#	src/test/java/testing/WallFollowAgentTest.java

S010MON reviewed (commented):
This PR does solve the issue, but it does so through a lot of hard coding the values using booleans, which isn't ideal

Jolijn2020 commented on src/main/java/app/model/TypeInformation.java:
```diff
@@ -0,0 +1,20 @@
+    {
+        // switch cases apparently don't support null,
+        // but wanted to keep switch-case cause cleaner and easier to add stuff
+        if(type == null)
```
if the ray doesn't hit any (visible) intersection, it has type null. It happens quite alot when the rays aren't very long and caused errors.

Jolijn2020 committed:
set default boundary values of agent

Jolijn2020 commented:
> This PR does solve the issue, but it does so through a lot of hard coding the values using booleans, which isn't ideal

would it be better to use TypeInformation to return whether it is it or not, or is that still using booleans? (even if not stored in Boundary itself)

Or would a totally different way be better? I can't think of any other way that doesn't somehow store which furniture is solid, visible, and transparent.

Jolijn2020 committed:
refacter to make visible, transparent and solid values an EnumType, BoundaryType

Jolijn2020 committed:
small fix, null-handler exception

Jolijn2020 committed:
Merge remote-tracking branch 'origin/master' into make_target_visible

# Conflicts:
#	src/main/java/app/model/agents/AgentType.java

Jolijn2020 committed:
merge fix

S010MON committed:
functional implementation of loops

S010MON committed:
convert comment to java doc

S010MON committed:
Merge remote-tracking branch 'origin/make_target_visible' into make_target_visible

S010MON committed:
merge changes and remove unused imports

S010MON committed:
remove unused import and add javadoc

S010MON committed:
style change

S010MON committed:
remove redundant comment

S010MON committed:
adds filtering for solid rays

S010MON commented on src/main/java/app/model/agents/AgentImp.java:
```diff
@@ -120,33 +121,16 @@ public void draw(GraphicsContext gc)
-    }
-
-    public Vector closestAgent(Type agentType)
+    public Vector closestTypeSeen(Type type)
```
I'm worried that this will return the closest type and not the closest AgentType, which means that we will always get a boundary returned if it's next to a wall.  This method is specifically for detecting the nearest agent in order to conduct state changes

Jolijn2020 replied:
AgentType isn't about being Intruder or Guard though, it's about whether it's Target, ACO, WF, etc. We always use ACO as guard and WF as intruder now, but that won't always be the case if we're going to add new agents that can be either.

Type, on the other hand, can only be Intruder or Guard for agents. The rays don't store AgentType either, just Type. So it won't be able to use the method in regards to vision, so not at all.
Unless we make changes to AgentType and store it in Ray too or make a certain AgentType always intruder/guard.

Jolijn2020 replied:
And if you give Type.INTRUDER or Type.GUARD (or any other type), it will only consider Rays with that specific type. All others don't get through the if-statement

S010MON committed:
indentation style

S010MON commented on src/main/java/app/model/agents/TargetAgent.java:
```diff
@@ -15,11 +15,10 @@ public class TargetAgent extends AgentImp
     private final boolean DEBUG = false;
 
-    public TargetAgent(Vector position, Vector direction, double radius, Type type, Ray targetRay, World world)
+    public TargetAgent(Vector position, Vector direction, double radius, Type type, Ray targetRay)
```
This constructor is getting pretty ugly ...

Jolijn2020 replied:
yeah, but TargetAgent needs the targetRay in order to function, so it needs to be a parameter.
I already deleted the World since it isn't used anywhere, but the targetRay has to stay.

S010MON replied:
But the TargetAgent can see the target in the view?  And will need to update the target reference each time anyway, so why not just set this internally in the constructor instead?

Jolijn2020 replied:
removed it

S010MON commented on src/main/java/app/model/boundary/BoundaryImp.java:
```diff
@@ -56,4 +63,19 @@ public Vector getTeleport()
         return null;
     }
+
+    public boolean isSolid()
```
I thought the point of the enum was to get rid of all of these boolean flags and methods.  Like having a `SOLID_TRANSPARENT` enum and `SOLID_VISIBLE` to make up the 9 possible options within the enum and swich on that, so I don't really understand why you have all three methods still (as only the `isSolid()` is used)

Jolijn2020 replied:
they are used in GraphicsEngine line 61 and 71.
But I agree they can be removed and use it on the type itself. I debated over it, but since it's cleaner when using the method, I kept them in. I'll remove them though.

S010MON reviewed (commented):
I still have a few concerns regarding the effect on the clarity for users of the new features (i.e. those making the agents)  I have added my own changes to the PR to help show what I mean

Jolijn2020 committed:
refacter, remove isSolid(), isTransparent(), and isVisible() from Boundary

Jolijn2020 committed:
Merge remote-tracking branch 'origin/make_target_visible' into make_target_visible

# Conflicts:
#	src/main/java/app/model/agents/AgentImp.java

Jolijn2020 committed:
refacter, renamed 2 variables

trav-d13 commented on src/test/java/testing/WallFollowAgentTest.java:
```diff
@@ -29,11 +29,13 @@
     Vector initialDirection = new Vector(0,1);
     double moveLen = 20.0;
-    WallFollowAgent agent = new WallFollowAgent(initialPosition, initialDirection, 1, Type.INTRUDER, moveLen);
+    //WallFollowAgent agent = new WallFollowAgent(initialPosition, initialDirection, 1, Type.INTRUDER, moveLen);
```
Remove comment please

Jolijn2020 committed:
Update WallFollowAgentTest.java

deleted commented out code

S010MON commented on src/main/java/app/controller/GameEngine.java:
```diff
@@ -137,7 +138,7 @@ private boolean legalMove(Vector start, Vector end)
         for (Boundary bdy : map.getBoundaries())
         {
-            if(bdy.isCrossed(start, end))
+            if(bdy.isCrossed(start, end) && BoundaryType.isSolid(bdy.getBoundaryType()) == true)
```
` BoundaryType.isSolid(bdy.getBoundaryType()) == true`

replace with 

`BoundaryType.isSolid(bdy.getBoundaryType())`

as isSolid() returns a boolean?

Jolijn2020 replied:
ah yes, that's kinda stupid

Jolijn2020 committed:
Merge remote-tracking branch 'origin/master' into make_target_visible

# Conflicts:
#	src/main/java/app/model/agents/WallFollow/WallFollowAgent.java

Jolijn2020 committed:
merging conflicts

Jolijn2020 committed:
Merge remote-tracking branch 'origin/master' into make_target_visible

# Conflicts:
#	src/main/java/app/controller/graphicsEngine/GraphicsEngine.java
#	src/main/java/app/model/Map.java
#	src/main/java/app/model/agents/ACO/AcoAgent.java
#	src/main/java/app/model/agents/AgentImp.java
#	src/main/java/app/model/agents/AgentType.java
#	src/main/java/app/model/agents/DirectionFollowAgent/DirectionFollowAgent.java
#	src/main/java/app/model/agents/TargetAgent.java
#	src/main/java/app/model/agents/WallFollow/WallFollowAgent.java
#	src/main/java/app/model/boundary/BoundaryFactory.java
#	src/main/java/app/model/boundary/BoundaryImp.java
#	src/main/java/app/model/boundary/VisibleBoundary.java

Jolijn2020 committed:
recover TargetAgent

Jolijn2020 committed:
merge fixes

Jolijn2020 committed:
fixed bug

Jolijn2020 committed:
Store Type in Boundary
This reduces complexity of having to go through all furniture to check if intersection is at that position

Jolijn2020 committed:
reduced complexity GraphicsEngine using Type stored in Boundary

Jolijn2020 committed:
removed TODO: comment

Jolijn2020 committed:
cleaned up code

Jolijn2020 committed:
deleted old code GraphicsEngine

Jolijn2020 committed:
filter for agents in noWallDetected in AgentImp and added type to Ray.toString

Jolijn2020 committed:
cleaned up code/comments

Jolijn2020 committed:
generalized isHit and isCrossed methods in BoundaryImp

Jolijn2020 committed:
fixed implementation of Rays with Type null

Jolijn2020 committed:
added Intruder and Guard types to BoundaryType.of(Type)

Jolijn2020 committed:
generalized BoundaryImp to limit needed class-extensions

Jolijn2020 committed:
detectCardinalPoint(), detect Closest solid intersection instead of first in list
